File: src/platform/transport/stream_router.py
# Requirement IDs: TRN-02, TRN-03, GOV-RES-01 | Backend route wiring (DP-B §4.2.1)
# ILLUSTRATIVE WIRING — shows how the FastAPI backend exposes the SSE stream at
# GET /events/stream with media_type text/event-stream. The Python publisher
# mirrors src/platform/transport/publisher.ts (same EventEnvelope contract,
# TRN-06 single source of truth); isolation per connection (TRN-RES-01) keeps a
# broken client from crashing the BE process (GOV-RES-01).
from __future__ import annotations

import asyncio
import json
import uuid
from datetime import datetime, timezone
from typing import Any, AsyncIterator

from fastapi import APIRouter, Request
from fastapi.responses import StreamingResponse

router = APIRouter()

_SSE_HEADERS = {"Cache-Control": "no-cache", "Connection": "keep-alive"}


class SseFanout:
    """Minimal SSE hub — one queue per connected client (TRN-RES-01)."""

    def __init__(self) -> None:
        self._queues: list[asyncio.Queue[dict[str, Any]]] = []

    def attach(self) -> asyncio.Queue[dict[str, Any]]:
        q: asyncio.Queue[dict[str, Any]] = asyncio.Queue()
        self._queues.append(q)
        return q

    def detach(self, q: asyncio.Queue[dict[str, Any]]) -> None:
        if q in self._queues:
            self._queues.remove(q)

    async def publish(self, envelope: dict[str, Any]) -> None:
        # allSettled-style isolation: a full/dead queue is dropped, never raised.
        for q in list(self._queues):
            try:
                q.put_nowait(envelope)
            except Exception:  # noqa: BLE001 - degrade safely (GOV-RES-01)
                self.detach(q)


HUB = SseFanout()


async def _stream(request: Request) -> AsyncIterator[str]:
    q = HUB.attach()
    try:
        yield ": connected\n\n"
        while True:
            if await request.is_disconnected():
                break
            envelope = await asyncio.wait_for(q.get(), timeout=15.0)
            yield f"id: {envelope['sequence']}\nevent: envelope\ndata: {json.dumps(envelope)}\n\n"
    except asyncio.TimeoutError:
        yield ": keepalive\n\n"  # comment frame as SSE heartbeat
    finally:
        HUB.detach(q)


@router.get("/events/stream")
async def events_stream(request: Request) -> StreamingResponse:
    return StreamingResponse(_stream(request), media_type="text/event-stream", headers=_SSE_HEADERS)


async def publish(step_id: str, status: str, payload: dict[str, Any] | None = None,
                  sequence: int = 0, trace_id: str | None = None) -> None:
    """Engine-facing helper mirroring Publisher.publish (§4.2.1)."""
    envelope: dict[str, Any] = {
        "step_id": step_id,
        "status": status,
        "payload": payload or {},
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "sequence": sequence,
        "trace_id": trace_id or str(uuid.uuid4()),
    }
    # RES-04 schema validation happens before emit in the real publisher; this
    # illustrative route only demonstrates the fanout attachment.
    await HUB.publish(envelope)

File: src/platform/transport/stream_router.ts
// Requirement IDs: TRN-02, TRN-03, GOV-RES-01 | Backend route wiring (DP-B §4.2.1)
// ILLUSTRATIVE WIRING — shows how a Node HTTP router exposes the SSE publisher
// at GET /events/stream (Express/Fastify equivalents attach identically via
// asSseStream). Application code imports only from src/platform/transport
// (GOV-REU-03); this file documents the server-side attachment pattern used by
// tests (TRN-AC-01/02). Uses structural http types so no framework dep is
// required in the platform layer.
import http from "node:http";
import type { IncomingMessage, Server, ServerResponse } from "node:http";
import { createPublisher } from "./index.js";
import type { CollectablePublisher } from "./index.js";

/**
 * Mount the event endpoints on an existing node:http request handler:
 *   GET /events/stream → SSE fanout (TRN-03)
 *   GET /events        → non-streaming fallback snapshot (TRN-RES-03, §4.6)
 * Switching TRANSPORT requires no call-site edit (GOV-REU-01): both routes are
 * served from the same transport-agnostic publisher facade.
 */
export function handleEventStream(
  publisher: CollectablePublisher,
): (req: IncomingMessage, res: ServerResponse) => void {
  // media_type text/event-stream equivalent for Node: headers are set by the
  // SSE adapter (Content-Type: text/event-stream, Cache-Control: no-cache).
  return (req, res) => {
    const url = (req.url ?? "").split("?")[0] ?? "";
    if (url === "/events") {
      // TRN-RES-03 synchronous final response — same envelope contract (TRN-06).
      const traceId = new URL(req.url ?? "/events", "http://localhost").searchParams.get("trace_id") ?? undefined;
      const snapshot = publisher.collect(traceId ?? undefined);
      if (!snapshot) {
        res.writeHead(404, { "Content-Type": "application/json" }).end(JSON.stringify({ error: "no events collected" }));
        return;
      }
      res.writeHead(200, { "Content-Type": "application/json; charset=utf-8", "Cache-Control": "no-cache" });
      res.end(JSON.stringify(snapshot));
      return;
    }
    if (!req.url || !req.url.startsWith("/events/stream")) {
      res.writeHead(404).end();
      return;
    }
    publisher.asSseStream(req, res); // one fanout connection per client (TRN-RES-01)
    req.on("close", () => console.debug("[events] client detached"));
  };
}

/** Example bootstrap (illustrative only — not executed by library consumers). */
export function createExampleServer(): Server {
  const publisher = createPublisher() as CollectablePublisher;
  return http.createServer(handleEventStream(publisher));
}
